keps/sig-cloud-provider/20190422-cloud-controller-manager-migration.md

```diff
@@ -0,0 +1,211 @@
+* Define a clear migration process for existing clusters to use the cloud-controller-manager instead of the kube-controller-manager for cluster integrations with any cloud provider.
+* The migration process should work for single node control planes and HA control planes with minimal risk to users.
+
+### Non-Goals
+
+* Removing cloud provider code from the core Kubernetes tree, this effort is separate and is covered in [KEP-removing-in-tree-providers](https://github.com/kubernetes/enhancements/blob/master/keps/sig-cloud-provider/20190125-removing-in-tree-providers.md)
+* Improving the scalability of controllers by running controllers across multiple components (with or without leader election).
+* Migrating cloud-based volume plugins to CSI. This is a separate effort led by SIG Storage. See [this proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md) for more details.
+
+## Proposal
+
+Today, the kube-controller-manager (KCM) and cloud-controller-manager (CCM) run independent of each other.
+This means that both the KCM or the CCM can run the cloud specific control loops for a given cluster.
+For a highly available control plane to migrate from running only the KCM to running both the KCM and
+the CCM requires that only one process in the control plane is running the cloud specific controllers.
+This becomes non-trivial when introducing the CCM that runs overlapping controllers with the KCM.
+
+For environments that can tolerate downtime, the control plane can be taken down in order to
+reconfigure components to use the CCM, and then bring the control plane back up. This ensures that only
+1 component can be running the set of cloud controllers. For environments that have stricter requirements
+for uptime, some level of coordination is required between the two components to ensure that upgrading
+control planes does not result in running the same controller in more than one place while also accounting for version skew.
+
+In order to coordinate the cloud-specific controllers across the KCM and the CCM, this KEP proposes a
+_primary_ and N configurable _secondary_ (a.k.a migration) leader election locks in the KCM and the CCM.
+The primary lock represents the current leader election resource lock in the KCM and the CCM. The set of
+secondary locks are defined by the cloud provider and run in parallel to the primary locks. For a migration
+lock defined by the cloud provider, the cloud provider also determines the set of controllers run within the
+migration lock and the controller manager it should run in - either the CCM or the KCM.
+
+The properties of the migration lock are:
+  * must have a unique name
+  * the set of controllers in the lock is immutable.
+  * no two migration locks should have overlapping controllers
+  * the controller manager where the lock runs can change across releases.
+  * can only run in either the kube-controller-manager or the cloud-controller-manager.
```

Comment:
Might be good to be a bit more explicit. For a minor release it should run exclusively in one type of controller manager. KCM or CCM.